### PR TITLE
feat(MeshHTTPRoute): initial plugin

### PR DESCRIPTION
**Pull request by michaelbeaumont**
This is only the plugin, the resource is extremely simplied and there is no extra validation.

### Checklist prior to review

- [x] Link to docs PR or issue --
- [x] Link to UI issue or PR --
- [x] Is the [issue worked on linked][1]? -- #5470 
- [x] The PR does not hardcode values that might break projects that depend on kuma (e.g. "kumahq" as a image registry) --
- [x] The PR will work for both Linux and Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Unit Tests --
- [x] E2E Tests -- none
- [x] Manual Universal Tests --
- [x] Manual Kubernetes Tests --
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [x] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin_test.go
Comment (jakubdyszkiewicz) [inline pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin_test.go:240]: nit: I'd use https://github.com/kumahq/kuma/blob/master/pkg/test/resources/builders/targetref_builder.go

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin_test.go
Comment (jakubdyszkiewicz) [inline pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin_test.go:308]: nit: `samples.MeshDefault()`

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:64]: I agree but i think this plugin should "just work" with other plugins, which I think depend on OriginOutbound

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:37]: Yeah, I think this was a bad refactoring of the configurers

Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:42]: fixed

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (jakubdyszkiewicz) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:69]: ah, good point. Other plugins have dependency on that

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:210]: I believe this refers to the virtual outbounds generated for cross-mesh gateway listeners

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go]: yeah maybe you're right. I did this on purpose but I conflated having this with the current `TrafficRoute` top level `destination`, which serves as a fallback. but I think it best to leave it up to the user to provide a fallback as a rule.

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin_test.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin_test.go:194]: I don't know, for some reason I settled on the first. Probably because the indentation got so bad somewhere. Why can't gofmt just be stricter... If we use [gofumpt](https://github.com/mvdan/gofumpt) it'd pick the second way.

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go]: done

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:43]: or rather for this PR I think i'll just keep the cluster cache thing

File: pkg/plugins/policies/meshhttproute/api/v1alpha1/meshhttproute.go
Comment (lahabana) [inline pkg/plugins/policies/meshhttproute/api/v1alpha1/meshhttproute.go:39]: There's no way to adapt the policy generation?

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (lahabana) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:42]: Shouldn't this `FindRoutes` be moved inside `MatchedPolicies` and `TypedMatchingPolicies` be extended with a `ToRouteRules` ? This would make extending the inspect API easier in the future.

**Review by lahabana — COMMENTED** [review]
I think reworking `MatchedPolicies` is required to be able to make the inspect API work in the future.

File: pkg/plugins/policies/meshhttproute/api/v1alpha1/meshhttproute.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/api/v1alpha1/meshhttproute.go:39]: Probably, but it's not worth refactoring IMO. This `Default` only shows up in the code, not the CRDs or serialized resources.

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin_test.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin_test.go:57]: https://github.com/kumahq/kuma/pull/5576

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:236]: I think it can be removed. It's used in `OutboundProxyGenerator` for conditionally applying rate limits but this will have to be done in the `MeshRateLimit` plugin.

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go]: It's a shame go doesn't allow for ergonomic data structures. We could use something like https://github.com/mitchellh/hashstructure so we don't have to implement this ourselves.

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go]: OK, I've made fixed it so that the default rule only takes effect if the user doesn't specify `backendRefs` for `Prefix /`

File: pkg/plugins/policies/meshhttproute/api/v1alpha1/zz_generated.resource.go
Comment (lobkovilya) [inline pkg/plugins/policies/meshhttproute/api/v1alpha1/zz_generated.resource.go:130]: This doesn't look right :D

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:27]: Yeah I'll tackle this in a separate PR cause I think "be the same as OutboundProxyGenerator" is a good goal for a first PR

File: pkg/plugins/policies/meshhttproute/api/v1alpha1/meshhttproute.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/api/v1alpha1/meshhttproute.go:39]: I think it may have been solved by adding `skip_getdefault`

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:47]: Hold on, the result of `FindRoutes` depends on which outbound we're talking about.
Right now `ToRules` contains an "unmerged" set of `RuleConfs`, but we can't merge it until we know which outbound we're talking about. This is the same in other plugins, in that we call `ToRules.Rules.Compute` for each outbound. Does the inspect API support seeing the rules for a specific outbound yet?

Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:47]: > This is indeed better for the future inspect API because the existing approach doesn't allow you to see the resulting RuleConf in TypeMatchingPolicies.

But I don't see why we should create a `ToRouteRules`, the problem is more that we can't put the merged rules into `ToRules`, isn't it?

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:145]: Perhaps it can be done generically by flagging a struct field, in this case `Rule.Matches` to be used a map key.

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:47]: I'd much rather make `ToRules` more flexible than add a special case

Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:47]: @lahabana @lobkovilya Can we tackle the inspect API in a separate PR? It's not broken ATM and this one is pretty big already

File: pkg/plugins/policies/meshhttproute/api/v1alpha1/zz_generated.resource.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/api/v1alpha1/zz_generated.resource.go:130]: https://github.com/kumahq/kuma/pull/5595 @lobkovilya

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin_test.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin_test.go:164]: @jakubdyszkiewicz I'm going to resolve this because IMO we shouldn't be policing `gofmt`ed code in reviews. If there are problems with `gofmt` accepted code, we should switch to something like `gofumpt`

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go]: @jakubdyszkiewicz does this look good?

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:27]: related to the following comment about cluster cache

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/plugin.go]: I'm going to leave this as is

**Comment by michaelbeaumont** [comment]
Outstanding: https://github.com/kumahq/kuma/issues/5470#issuecomment-1375558476

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (jakubdyszkiewicz) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go]: Ok, but please make a note to change it before the 2.1 release. Thank you

Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go]: done, added to the issue https://github.com/kumahq/kuma/issues/5470#issuecomment-1375558476

**Review by jakubdyszkiewicz — APPROVED** [review]

**Review by lobkovilya — APPROVED** [review]

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:145]: @lobkovilya for this specifically, I created https://github.com/michaelbeaumont/kuma/commit/65711a72cf21f98a290ee36eeb50eaec6822c3cc#diff-a6760caedf3999fbc634e9fca1c2653b9bb1218d7df3e9bc7ef0cdfc9112bf3c to do it generically

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go
Comment (lahabana) [reply pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go:47]: Ok as long as it's a fast follow :)

**Pull request merged**
